Added flag to CollectionChangeSet to indicate cleared collection

papafe opened the pull request:
Added a new flag to `CollectionChangeSet` to indicate when `remove_all` is called on collections. 

## ☑️ ToDos
* [x] 📝 Changelog update
* [x] 🚦 Tests

fealebenpae reviewed (approved):
Looks good. Remember to add a changelog entry.

tgoyne commented:
`merge()` needs to propagate `collection_was_cleared` (and there needs to be tests for that). The current tests would pass if `collection_was_cleared` was unconditionally set to `true`.

papafe commented:
@tgoyne I have propagated the flag in `merge()`, hoping that my understanding is correct. I also added some more tests

tgoyne commented on src/realm/object-store/impl/collection_change_builder.cpp:
```diff
@@ -147,6 +149,8 @@ void CollectionChangeBuilder::merge(CollectionChangeBuilder&& c)
         collection_root_was_deleted = true;
     }
 
+    collection_was_cleared = c.collection_was_cleared;
```
This results is inconsistent behavior. If a clear is followed by an insert in a single write transaction it'll report `collection_was_cleared=true` plus an insertion, but if they're in two separate transactions that get coalesced it'll report `collection_was_cleared=false` plus an insertion. Whether or not this should stay set if there's subsequent modifications comes down to what the SDK needs, but it does need to behave the same way in both cases.

papafe replied:
I think that in this case what counts is the final state of the collection, so in this case probably having `collection_was_cleared=false` is the right choice.

nirinchev replied:
Agree - I think we should only report `collection_was_cleared` if the final state of the collection is empty. So an insertion should reset the `collection_was_cleared` flag to `false`. Might be a good idea to add an explicit test for that.

tgoyne commented on src/realm/object-store/impl/collection_change_builder.cpp:
```diff
@@ -26,12 +26,14 @@ using namespace realm;
 
 CollectionChangeBuilder::CollectionChangeBuilder(IndexSet deletions, IndexSet insertions, IndexSet modifications,
-                                                 std::vector<Move> moves, bool root_was_deleted)
+                                                 std::vector<Move> moves, bool collection_was_cleared,
```
Adding the new argument in this position means that any code which was calling this constructor and passing in a bool will still compile, but now do something different.

papafe commented:
@tgoyne I am setting `collection_was_cleared=false` when something is added to the collection, so that the behaviour should be consistent with the final state of the collection. I've added also two tests to verify that this works. 
I've also changed the position of the new boolean, so that it appears after the one that was already there to avoid issues.
